### ISSUE TITLE
Integrate connection to portal into tunnel

### ISSUE THREAD

**Issue by thomaseizinger** [firezone/firezone#3837]
Once https://github.com/firezone/firezone/pull/3682 is merged, the events emitted by `firezone_tunnel` are consumed within an eventloop for both client and gateway.

We can massively simplify this by moving this eventloop into the `firezone_tunnel` crate:

- Pretty much all the public API of `firezone_tunnel::Tunnel` will go away
- No more need for the `connlib_shared` crate, all the messages will just be defined within `firezone_tunnel`
- Less indirection: We can directly send a message to the portal upon certain events in the tunnel, no more need to explicitly map to other events
- If #3830 pans out as planned, a `Session` is just a `Tunnel` which means we can remove the concept of a `Session`

Overall, I expect this to be a big net-positive in terms of codesize (i.e. lots of deleted lines).

**Comment by thomaseizinger** [comment]
Once this is in place, Rust-based clients like the Tauri GUI and the CLI client can just use the `Tunnel` directly and spawn their own task onto their runtime, which should remove a lot of worker threads.

**Comment by jamilbk** [comment]
@thomaseizinger Still relevant?

**Comment by thomaseizinger** [comment]
> @thomaseizinger Still relevant?

Yes.

**Comment by thomaseizinger** [comment]
If we temporarily duplicate the `Tunnel` struct, we should be able to do this for client and gateway separately and thus reduce the complexity of code that needs to be touched at once.

**Comment by thomaseizinger** [comment]
As a first step towards this, we can move the call to `PhoenixChannel::new` into the respective clients. That should reduce the number of things we have to pass down in `ConnectArgs`.

**Comment by thomaseizinger** [comment]
I just had another crack at this and it is still quite a large chunk of code that would move at once and it would need some hacks. It turns out that we are still doing quite a bit of mapping back and forth between events from the tunnel and messages to send via the portal connection.

I think a smaller, useful step would be to move the messages into `firezone_tunnel` and use those as the API with the respective event loops.

**Comment by jamilbk** [comment]
Is this still relevant?

**Comment by thomaseizinger** [comment]
I think it would still be worth doing. We can still have a separate task for the connection, similar to how we spawn threads for the sockets from within `firezone_tunnel`. This would make the event-loops for both client and gateway smaller so less of this logic is split across the codebase.

**Comment by thomaseizinger** [comment]
But if you feel like tidying up, we can close this. Doing this is kind of blocked on releasing Gateway 1.5 because of all the legacy code that we have inside the Gateway's event-loop which doesn't map very nicely to what we do in the tunnel.
